port-target-cache not available with pytest-xdist

https://github.com/espressif/esp-bsp/actions/runs/14508751792/job/40703558008

```
[gw10] linux -- Python 3.11.12 /usr/local/bin/python3.11
/usr/local/lib/python3.11/site-packages/pytest_embedded/plugin.py:630: in port_target_cache
    with shelve.open(_cache_file_path) as f:
/usr/local/lib/python3.11/shelve.py:243: in open
    return DbfilenameShelf(filename, flag, protocol, writeback)
/usr/local/lib/python3.11/shelve.py:227: in __init__
    Shelf.__init__(self, dbm.open(filename, flag), protocol, writeback)
/usr/local/lib/python3.11/dbm/__init__.py:95: in open
    return mod.open(file, flag, mode)
E   _gdbm.error: [Errno 11] Resource temporarily unavailable: '/tmp/pytest-embedded/pytest-embedded-cache/port_target_cache'

During handling of the above exception, another exception occurred:
/usr/local/lib/python3.11/site-packages/pytest_embedded/plugin.py:[63](https://github.com/espressif/esp-bsp/actions/runs/14508751792/job/40703558008#step:8:64)3: in port_target_cache
    os.remove(_cache_file_path)
E   FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-embedded/pytest-embedded-cache/port_target_cache'
```

For now, the `port_target_cache` is held for the entire process. Let's change it to be read and written on-demand, using a file lock. Since `pytest-xdist` runs tests with multiprocessing, standard `multiprocessing.Lock` or `threading.Lock` won't work.

--- a/pytest-embedded/pyproject.toml
+++ b/pytest-embedded/pyproject.toml
@@ -31,6 +31,7 @@ requires-python = ">=3.7"
 dependencies = [
     "pytest>=7.0",
     "pexpect>=4.4",
+    "filelock>=3.12.2"
 ]
 
 [project.urls]

--- a/pytest-embedded/pytest_embedded/plugin.py
+++ b/pytest-embedded/pytest_embedded/plugin.py
@@ -17,6 +17,7 @@ import xml.dom.minidom
 from collections import Counter
 from operator import itemgetter
 
+import filelock
 import pytest
 from _pytest.config import Config
 from _pytest.fixtures import (
@@ -625,18 +626,20 @@ def cache_dir(request: FixtureRequest) -> str:
 def port_target_cache(cache_dir) -> t.Dict[str, str]:
     """Session scoped port-target cache, for esp only"""
     _cache_file_path = os.path.join(cache_dir, 'port_target_cache')
+    lock = filelock.FileLock(f'{_cache_file_path}.lock')
     resp: t.Dict[str, str] = {}
-    try:
-        with shelve.open(_cache_file_path) as f:
-            resp = dict(f)
-    except dbm.error:
-        os.remove(_cache_file_path)
+    with lock:
+        try:
+            with shelve.open(_cache_file_path) as f:
+                resp = dict(f)
+        except dbm.error:
+            os.remove(_cache_file_path)
 
     yield resp
-
-    with shelve.open(_cache_file_path) as f:
-        for k, v in resp.items():
-            f[k] = v
+    with lock:
+        with shelve.open(_cache_file_path) as f:
+            for k, v in resp.items():
+                f[k] = v
 
 
 @pytest.fixture(scope='session')

